File: DFS_Of_Graph.java


class Solution {
    // Function to return a list containing the DFS traversal of the graph.
    public void dfs(ArrayList<ArrayList<Integer>> adj, int vis[], ArrayList<Integer> a,int start)
    {
        a.add(start);
        vis[start]=1;
        for( int x: adj.get(start))
        {
            if(vis[x]==0)
            {
                vis[x]=1;
                dfs(adj,vis,a,x);
            }
        }
    }
    public ArrayList<Integer> dfsOfGraph(int V, ArrayList<ArrayList<Integer>> adj) {
        // Code here
        ArrayList<Integer> a=new ArrayList<Integer>();
        int vis[]=new int[V+1];
        dfs(adj,vis,a,0);
        return a;
    }
}

File: Difficulty: Easy/DFS of Graph/dfs-of-graph.py
#User function Template for python3

class Solution:
    
    #Function to return a list containing the DFS traversal of the graph.
    def dfs(self, adj):
        # code here
        v=set()
        l=[]
        def f(n) :
            l.append(n)
            for x in adj[n] :
                if x not in v :
                    v.add(x)
                    f(x)
        for i in range(len(adj)) :
            if i not in v :
                v.add(i)
                f(i)
        
        return l

#{ 
 # Driver Code Starts
import sys
#Position this line where user code will be pasted.


def main():
    tc = int(sys.stdin.readline().strip())

    for _ in range(tc):
        V = int(sys.stdin.readline().strip())
        adj = []

        for _ in range(V):
            input_line = sys.stdin.readline().strip()
            node = list(map(int, input_line.split())) if input_line else []
            adj.append(node)

        obj = Solution()
        ans = obj.dfs(adj)
        print(" ".join(map(str, ans)))
        print("~")


if __name__ == "__main__":
    main()

# } Driver Code Ends